### PR TITLE
Implement isolated inference - phase II

### LINKED ISSUE #31112: Isolated inference for object `init` method should consider initializers of fields
**Description:**
$title.

**Steps to reproduce:**
The following prints true atm.

```ballerina
int[] & readonly nonFinalVar = [];

class Foo {
    private final int[] & readonly x = nonFinalVar;

    function init() {
    }
}

function func() {
    Foo f = new;
}

public function main() {
    io:println(func is isolated function ());
}
```

But if `func` and therefore `Foo.init` is explicitly marked as `isolated`, the initializer will result in a compilation error.

```ballerina
int[] & readonly nonFinalVar = [];

class Foo {
    private final int[] & readonly x = nonFinalVar; // error: invalid access of mutable storage in the initializer of an object field

    isolated function init() {
    }
}

isolated function func() {
    Foo f = new;
}
```

**Affected Versions:**
slbeta1

### LINKED ISSUE #31627: Incorrect isolated inference with parameters annotated as `@isolatedParam`
**Description:**

**Steps to reproduce:**
The `f` method is not isolated (based on the current rules), but is currently inferred to be so.

```ballerina
service class Foo {
    remote function f() {
        int[] x = [1, 2];
        stream<int> y = x.toStream();

        int[] z = [];
        x.forEach(i => z.push(i));
    }
}
```

**Affected Versions:**
slbeta2

### PR DESCRIPTION
**Pull request by MaryamZi**
## Purpose
Implement isolated inference for variables and classes.

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/26045

This PR also
- fixes #31627
- fixes #31112

## Remarks
https://github.com/ballerina-platform/ballerina-lang/pull/29922

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java [hunk 1]
```diff
@@ -2235,6 +2360,17 @@ private boolean isObjectFieldDefaultValueRequiringIsolation(SymbolEnv env) {
         return isIsolated(initFunction.symbol.flags);
     }
 
+    private boolean isObjectFieldDefaultValue(SymbolEnv env) {
+        BLangNode node = env.node;
+        NodeKind kind = node.getKind();
+
+        if (kind != NodeKind.CLASS_DEFN) {
```
[the remark this answers is not in the document]
Comment (pcnfernando) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java:2367]: Also shall we simplify to kind == NodeKind.CLASS_DEFN

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java [hunk 2]
```diff
@@ -2932,11 +3122,15 @@ private boolean isReferenceToVarDefinedInSameInvokable(BSymbol currentOwner, BIn
     private boolean isIsolationInferableFunction(BLangFunction funcNode) {
         Set<Flag> flagSet = funcNode.flagSet;
 
-        if (flagSet.contains(Flag.PUBLIC) || flagSet.contains(Flag.INTERFACE)) {
+        if (flagSet.contains(Flag.INTERFACE)) {
             return false;
         }
 
         if (!flagSet.contains(Flag.ATTACHED)) {
+            if (flagSet.contains(Flag.PUBLIC)) {
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java:3130]: Shall we simplify to `return !flagSet.contains(Flag.PUBLIC);`

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java [hunk 3]
```diff
@@ -375,14 +413,17 @@ public void visit(BLangFunction funcNode) {
 
         analyzeNode(funcNode.body, funcEnv);
 
-        if (!isIsolated(funcNode.symbol.flags) && this.inferredIsolated && !
-                Symbols.isFlagOn(funcNode.symbol.flags, Flags.WORKER)) {
+        if (!isIsolated(symbol.flags) && this.inferredIsolated && !
+                Symbols.isFlagOn(symbol.flags, Flags.WORKER)) {
```
[the remark this answers is not in the document]
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java]: Merged this into one line and fixed the rest also, thanks!

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java [hunk 4]
```diff
@@ -2613,12 +2790,49 @@ private boolean isIsolatedExpression(BLangExpression expression, boolean logErro
...
+            return false;
+        }
+
+        if (tag != TypeTags.UNION) {
+            return false;
+        }
+
+        for (BType memberType : ((BUnionType) type).getMemberTypes()) {
+            if (!isSubTypeOfReadOnlyOrIsolatedObjectUnionWithInference(publiclyExposedObjectTypes, classDefinitions,
+                                                                       inferring, memberType, unresolvedSymbols)) {
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java:2829]: inferring will be anyway true here

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java [hunk 5]
```diff
@@ -2165,48 +2297,77 @@ private void analyzeRestArgsForRestParam(List<BLangExpression> restArgs, BInvoka
...
         if (lastArgIsVarArg) {
-            analyzeVarArgIsolatedness((BLangRestArgsExpression) lastArg, lastArg.pos);
+            analyzeVarArgIsolatedness(invocationExpr, (BLangRestArgsExpression) lastArg, lastArg.pos, expectsIsolation);
         }
     }
 
-    private void analyzeVarArgIsolatedness(BLangRestArgsExpression restArgsExpression, Location pos) {
+    private void analyzeVarArgIsolatedness(BLangInvocation invocationExpr, BLangRestArgsExpression restArgsExpression,
+                                           Location pos, boolean expectsIsolation) {
+        BInvokableSymbol symbol = (BInvokableSymbol) invocationExpr.symbol;
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java:2311]: Symbol not used

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java [hunk 6]
```diff
@@ -2010,9 +2144,8 @@ private void analyzeArgIsolatedness(List<BLangExpression> requiredArgs, List<BLa
 
             analyzeAndSetArrowFuncFlagForIsolatedParamArg(arg);
 
-            if (!Symbols.isFlagOn(arg.getBType().flags, Flags.ISOLATED)) {
-                dlog.error(arg.pos, DiagnosticErrorCode.INVALID_NON_ISOLATED_FUNCTION_AS_ARGUMENT);
-            }
+            handleNonExplicitlyIsolatedArgForIsolatedParam(invocationExpr, arg, expectsIsolation,
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java:2147]: Shall we manually check whether these lines are actually not covered?

**Review by pcnfernando — APPROVED** [review]
Tested. LGTM

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java (hunk 6 above)
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java:2147]: This PR just refactored the check into a method but seems like these weren't covered. Added tests specifically for this line (a334b49) but the rest seems to be blocked by https://github.com/ballerina-platform/ballerina-lang/issues/31847. Created an issue to track adding tests once that is fixed - https://github.com/ballerina-platform/ballerina-lang/issues/31848.

**Review by pcnfernando — APPROVED** [review]

**Pull request merged**